File: go-core/jsx_test.go
package main

import (
	"encoding/json"
	"math"
	"strings"
	"testing"
)

// Adversarial coverage for the JSX interpolation helpers (jsx.go). These are
// the security chokepoints: every caller-controlled string that lands in
// emitted ExtendScript routes through jsLit, so a break-out here is a script
// injection into the running Photoshop session. The tests assert that
// hostile inputs come back as inert, valid, quoted JS string literals that
// cannot escape their quotes — and that jsNum never emits a bare
// non-finite identifier.

// jsLit's contract is "JSON.stringify(String(value))": the output must be a
// valid JSON/JS string literal (parseable by encoding/json back to the exact
// input) that begins and ends with a double quote.

func assertInertLiteral(t *testing.T, in string) {
	t.Helper()
	out := jsLit(in)

	// 1. Must be wrapped in double quotes — never a bare token.
	if len(out) < 2 || out[0] != '"' || out[len(out)-1] != '"' {
		t.Fatalf("jsLit(%q) = %q: not a double-quoted literal", in, out)
	}

	// 2. Must round-trip back to the original string through a strict JSON
	//    parse. If a break-out char leaked unescaped, json.Unmarshal either
	//    fails or yields a different string.
	var back string
	if err := json.Unmarshal([]byte(out), &back); err != nil {
		t.Fatalf("jsLit(%q) = %q: not valid JSON literal: %v", in, out, err)
	}
	if back != in {
		t.Fatalf("jsLit(%q) round-tripped to %q, want %q", in, back, in)
	}

	// 3. The interior (literal minus the outer quotes) must contain no raw
	//    (unescaped) double quote — that is exactly the break-out vector.
	interior := out[1 : len(out)-1]
	for i := 0; i < len(interior); i++ {
		if interior[i] == '"' {
			// A quote is only legal if preceded by an odd run of backslashes.
			bs := 0
			for j := i - 1; j >= 0 && interior[j] == '\\'; j-- {
				bs++
			}
			if bs%2 == 0 {
				t.Fatalf("jsLit(%q) = %q: unescaped quote at %d (break-out)", in, out, i)
			}
		}
	}

	// 4. No literal newline or carriage return may survive — they'd terminate
	//    a line and could split the surrounding statement.
	if strings.ContainsAny(interior, "\n\r") {
		t.Fatalf("jsLit(%q) = %q: contains a raw newline/CR", in, out)
	}
}

func TestJsLitBreakOutAttempts(t *testing.T) {
	cases := []string{
		`a"); evil(("`,          // close-string + inject call
		`"`,                     // bare double quote
		`\`,                     // bare backslash
		`\"`,                    // escaped-quote sequence
		`\\"`,                   // backslash then quote
		"line1\nline2",          // embedded newline
		"tab\there",             // embedded tab
		"carriage\rreturn",      // embedded CR
		"</script>",             // HTML/script terminator
		"<!-- comment -->",      // HTML comment
		"a & b < c > d",         // raw &, <, > (must pass through, not HTML-escaped)
		"'; alert('x')",         // single-quote break-out
		"`${danger}`",           // template-literal injection
		"unicode: é中\U0001F600", // accented, CJK, emoji
		"null\x00byte",          // embedded NUL
		"",                      // empty string
		"plain layer name",      // benign control
		"Layer \"copy\" 2",      // benign embedded quotes
	}
	for _, in := range cases {
		assertInertLiteral(t, in)
	}
}

// jsLit must emit pure printable ASCII. The .jsx it feeds is written UTF-8
// with no BOM and no #encoding directive, so ExtendScript decodes it by the
// platform codepage: a raw 'ü' arrives mojibake and every comparison against
// it misses — which is what made deleting a German Photoshop's own
// 'Farbfüllung 1' fail. \uXXXX survives any codepage and the parser restores
// the exact character, so the round-trip assertion in assertInertLiteral is
// what proves the escape is lossless.
func TestJsLitEscapesNonASCII(t *testing.T) {
	cases := []struct{ in, want string }{
		{"Farbfüllung 1", "\"Farbf\\u00fcllung 1\""},
		{"Ebene 1", "\"Ebene 1\""}, // benign control: untouched
		{"中", "\"\\u4e2d\""},
		{"\U0001F3A8", "\"\\ud83c\\udfa8\""}, // astral → surrogate pair
	}
	for _, c := range cases {
		if got := jsLit(c.in); got != c.want {
			t.Errorf("jsLit(%q) = %s, want %s", c.in, got, c.want)
		}
		assertInertLiteral(t, c.in)
	}
}

// Belt and braces on the property the fix actually depends on: whatever the
// input, nothing outside printable ASCII may survive into the emitted literal.
func TestJsLitOutputIsAlwaysPrintableASCII(t *testing.T) {
	for _, in := range []string{
		"Farbfüllung 1", "é中\U0001F600", "a\u2028b", "naïve — dash", "\u007f",
	} {
		out := jsLit(in)
		for i := 0; i < len(out); i++ {
			if out[i] < ' ' || out[i] > '~' {
				t.Fatalf("jsLit(%q) = %q: byte %d (%#x) is outside printable ASCII", in, out, i, out[i])
			}
		}
	}
}

// Boundary cases for the ASCII escape, kept separate so each one names what it
// is guarding. DEL and the separators are the characters most likely to
// diverge from the TypeScript twin; the quote/backslash case proves the escape
// runs AFTER the JSON encoder and cannot double-process what it already did.
func TestJsLitEscapeBoundaries(t *testing.T) {
	cases := []struct{ name, in, want string }{
		{"DEL", "\u007f", "\"\\u007f\""},
		{"line separator", "\u2028", "\"\\u2028\""},
		{"paragraph separator", "\u2029", "\"\\u2029\""},
		{"highest code point", "\U0010FFFF", "\"\\udbff\\udfff\""},
		{"tilde stays raw", "~", "\"~\""},
		{"space stays raw", " ", "\" \""},
		{"already-escaped quote and backslash", "a\"b\\c", "\"a\\\"b\\\\c\""},
	}
	for _, c := range cases {
		if got := jsLit(c.in); got != c.want {
			t.Errorf("%s: jsLit(%q) = %s, want %s", c.name, c.in, got, c.want)
		}
		assertInertLiteral(t, c.in)
	}
}

// jsLit must NOT HTML-escape <, >, & — it mirrors JS JSON.stringify, which
// leaves them literal (the Go default would escape them and diverge from the
// TS output). Pin that explicitly.
func TestJsLitDoesNotHTMLEscape(t *testing.T) {
	out := jsLit("a & b < c > d")
	// Exact-equality pins it fully: the literal ampersand/lt/gt pass through
	// (Go's encoding/json default would emit unicode escapes for them);
	// SetEscapeHTML(false) is what keeps them literal and matches JS output.
	if out != `"a & b < c > d"` {
		t.Fatalf("jsLit HTML-escaped (want literal): got %q, want %q", out, `"a & b < c > d"`)
	}
}

// jsLit never returns the empty string (an empty value would be an unquoted,
// break-out-able blank slot). The empty-input case must still yield `""`.
func TestJsLitNeverEmpty(t *testing.T) {
	if got := jsLit(""); got != `""` {
		t.Fatalf("jsLit(\"\") = %q, want %q", got, `""`)
	}
}

// jsNum: strconv.FormatFloat renders NaN/Inf as the bare words NaN/+Inf/-Inf,
// which are NOT valid JS numeric literals. The guard must coerce any
// non-finite input to a finite literal so the emitted JSX never carries a bare
// identifier in a numeric slot.
func TestJsNumNonFinite(t *testing.T) {
	nonFinite := []float64{
		math.NaN(),
		math.Inf(1),
		math.Inf(-1),
	}
	for _, v := range nonFinite {
		out := jsNum(v)
		// Must not be any of the bare-identifier forms FormatFloat would emit.
		for _, bad := range []string{"NaN", "Inf", "+Inf", "-Inf", "Infinity", "+Infinity", "-Infinity"} {
			if out == bad {
				t.Fatalf("jsNum(%v) = %q: emitted bare non-finite identifier", v, out)
			}
		}
		// The coerced value must parse as a finite JS/JSON number.
		var n float64
		if err := json.Unmarshal([]byte(out), &n); err != nil {
			t.Fatalf("jsNum(%v) = %q: not a valid numeric literal: %v", v, out, err)
		}
		if math.IsNaN(n) || math.IsInf(n, 0) {
			t.Fatalf("jsNum(%v) = %q parsed back to a non-finite number", v, out)
		}
	}
}

func TestJsNumFinite(t *testing.T) {
	cases := []struct {
		in   float64
		want string
	}{
		{0, "0"},
		{1, "1"},
		{-5, "-5"},
		{2.5, "2.5"},
		{100, "100"},
		{0.01, "0.01"},
	}
	for _, c := range cases {
		if got := jsNum(c.in); got != c.want {
			t.Errorf("jsNum(%v) = %q, want %q", c.in, got, c.want)
		}
	}
}

func TestJsBool(t *testing.T) {
	if got := jsBool(true); got != "true" {
		t.Errorf("jsBool(true) = %q, want %q", got, "true")
	}
	if got := jsBool(false); got != "false" {
		t.Errorf("jsBool(false) = %q, want %q", got, "false")
	}
}

File: tests/unit/jsx.test.ts
/**
 * Tests for the JSX-interpolation escape helpers.
 *
 * Every ExtendScript snippet built by Editmamei depends on `jsLit` /
 * `jsNum` / `jsBool` to safely interpolate values from tool args. These
 * helpers are the injection-prevention boundary — adversarial inputs
 * (embedded quotes, line/paragraph separators, NaN coercion, BigInt,
 * Symbol, etc.) MUST land as valid ExtendScript literals or fail
 * deterministically. The indirect coverage via tests/unit/extendscript
 * doesn't enumerate these cases.
 */
import { describe, it, expect } from 'vitest';
import { jsLit, jsNum, jsBool } from '@editmamei/utils/jsx.js';

describe('jsLit', () => {
  it('wraps a plain string in quotes', () => {
    expect(jsLit('hello')).toBe('"hello"');
  });

  it('escapes embedded double quotes', () => {
    expect(jsLit('foo"bar')).toBe('"foo\\"bar"');
  });

  it('escapes backslashes', () => {
    expect(jsLit('C:\\path')).toBe('"C:\\\\path"');
  });

  it('escapes newlines and tabs', () => {
    expect(jsLit('a\nb\tc')).toBe('"a\\nb\\tc"');
  });

  it('escapes carriage returns', () => {
    expect(jsLit('a\rb')).toBe('"a\\rb"');
  });

  it('preserves the empty string as ""', () => {
    expect(jsLit('')).toBe('""');
  });

  it('coerces non-strings via String()', () => {
    expect(jsLit(42)).toBe('"42"');
    expect(jsLit(null)).toBe('"null"');
    expect(jsLit(undefined)).toBe('"undefined"');
    expect(jsLit(true)).toBe('"true"');
  });

  it('round-trips line/paragraph separators through JSON.parse', () => {
    // U+2028 / U+2029 are valid string-literal characters in ECMAScript
    // 2019+ \u2014 both Node JS and ExtendScript accept them inside string
    // literals. JSON.stringify is allowed to emit them either as raw
    // chars or as \\u escapes; the contract we care about is that the
    // literal round-trips and that interpolation can't break out of the
    // string.
    const LS = '\u2028';
    const PS = '\u2029';
    const input = `a${LS}b${PS}c`;
    const out = jsLit(input);
    expect(JSON.parse(out)).toBe(input);
  });

  it('handles NUL bytes via the standard JSON escape', () => {
    const out = jsLit('a\x00b');
    expect(JSON.parse(out)).toBe('a\x00b');
  });

  it('handles long strings without truncation', () => {
    const longStr = 'x'.repeat(10_000);
    const out = jsLit(longStr);
    expect(JSON.parse(out)).toBe(longStr);
  });
});

describe('jsNum', () => {
  it('passes finite numbers through', () => {
    expect(jsNum(3.14, 0)).toBe('3.14');
    expect(jsNum(-7, 0)).toBe('-7');
    expect(jsNum(0, 99)).toBe('0');
  });

  it('coerces numeric strings', () => {
    expect(jsNum('42', 0)).toBe('42');
  });

  it('falls back on NaN', () => {
    expect(jsNum(NaN, 99)).toBe('99');
    expect(jsNum('not a number', 12)).toBe('12');
  });

  it('falls back on Infinity / -Infinity', () => {
    expect(jsNum(Infinity, 5)).toBe('5');
    expect(jsNum(-Infinity, 5)).toBe('5');
  });

  it('coerces / falls back per Number()-cast semantics', () => {
    // jsNum's contract is "finite number passes through, else fallback."
    // The cast-to-number step lands per JS Number() semantics: Number(null)
    // is 0 and Number([]) is 0 — both finite, both pass through as '0'.
    // Number(undefined) and Number({}) are NaN — both hit the fallback.
    expect(jsNum(null, 1)).toBe('0');
    expect(jsNum([], 1)).toBe('0');
    expect(jsNum(undefined, 1)).toBe('1');
    expect(jsNum({}, 1)).toBe('1');
  });

  it('never returns a value with embedded characters that would break JSX', () => {
    // Synthesize a few adversarial inputs and verify the output is
    // a plain numeric literal.
    const bad = ['1; alert(1)', '1)//hack', '"; x = 1; "'];
    for (const input of bad) {
      const out = jsNum(input, 0);
      expect(out).toMatch(/^-?\d+(\.\d+)?$/);
    }
  });
});

describe('jsBool', () => {
  it('emits the canonical boolean literal', () => {
    expect(jsBool(true, false)).toBe('true');
    expect(jsBool(false, true)).toBe('false');
  });

  it('coerces "true" / "false" strings', () => {
    expect(jsBool('true', false)).toBe('true');
    expect(jsBool('false', true)).toBe('false');
  });

  it('falls back for any other value', () => {
    expect(jsBool(undefined, true)).toBe('true');
    expect(jsBool(null, false)).toBe('false');
    expect(jsBool(0, false)).toBe('false');
    expect(jsBool(1, false)).toBe('false');
    expect(jsBool('yes', false)).toBe('false');
    expect(jsBool({}, true)).toBe('true');
  });
});
